[1.20.1] Seal events that only have `@ApiStatus.Internal` constructors

PaintNinja opened the pull request:
Adding `final` to these event classes allows for additional optimisations to be made by both the JVM and EventBus (such as [EventBus#85](<https://github.com/MinecraftForge/EventBus/pull/85>)). Partial backport of #10515 to 1.20.1.

It's not possible to extend these events without calling the constructor that is marked as internal, therefore extending was considered for internal use only and should be made `final`.

~~`sealed` has been skipped because EventBus 6 doesn't support using this for optimisations yet, although I can do that as well if wanted? Increases the chances of breaking a mod that relied on internals, but would mean I don't have to make another PR adding sealed later if I figure out a way for EB6 to recognise it.~~ [EventBus#98](<https://github.com/MinecraftForge/EventBus/pull/98>) adds support for optimisations on `sealed`.

Sidenote: it's a shame that so many events in 1.20.1 didn't have their constructors marked as internal by the old team before the first RB, especially the tick events... :/

PaintNinja committed:
Seal events that only have `@ApiStatus.Internal` constructors

Adding `final` to these event classes allows for additional optimisations to be made by both the JVM and EventBus (such as EventBus#85). Partial backport of #10515.

It's not possible to extend these events without calling the constructor that is marked as internal, therefore extending was considered for internal use only and should be made `final`.

Jonathing commented:
Yes, they did mark these events as internal. But I'm still going to need to test this on all the kitchen slop packs since it's entirely possible that there could be events that inherit them. I would prefer not to have a repeat of 47.4.10.

Jonathing commented:
Assigning myself to this issue as I am blocking its merge until I've tested it.

KnightMiner commented:
Is there a practical benefit to this change? Seems like a breaking change purely for the sake of cleanup; risks breaking mods without any actual advantage.

It's reasonable to do during a breaking change window on a new version, but not on an LTS version.

Jonathing commented:
I understand why Paint made this. These events are marked as internal, after all. But I am all too aware that implementations do not have to respect those annotations.

The fact that you've commented on this shows that it really won't be worth merging this.

PaintNinja commented:
Yes, it's explained in the two PRs I linked in this PR's description. But in a nutshell, it speeds up event posting by about 33% for events that are known to not be cancellable, but to know that the event needs to be sealed as otherwise it's technically possible for a subclass of the event to be cancellable, which in turn causes the superclass to be cancellable.

There's also the argument that it helps prevent unintended breaking changes, as adding new methods to an existing event with an internal constructor could break subevents with the same method name. This is not merely a cleanup.

This shouldn't really be a breaking change because unlike the 1.21.5 PR, this one only applies to events that already had their constructors marked as internal in 47.1.0. Stability guarantees aren't made for things that are explicitly marked as internal, but as Jonathan said we'll test it with a large pack anyway just in case.

If you are aware of mods that would break as a result of this, please let us know and we can remove the internal annotation for the affected events and intentionally support it such as by making some fields protected instead of private too.

PaintNinja committed:
Add sealed to abstract events with only internal constructors

PaintNinja commented:
Added `sealed` where appropriate following some discussion on Discord: https://discord.com/channels/1129059589325852724/1129095235889270844/1427249722074206238

Jonathing commented:
We can't merge this as it currently stands. I've tested against Better Minecraft 4 and All of Create, and they both crash on startup due to a mod attempting to inherit from RenderTooltipEvent. This proves that modders do not care about internal annotations, and that those implementations are already permeating several modpacks in the wild that still update Forge. I don't think the performance gain this offers is worth the damage it could bring to popular mod(pack)s, even if the mods in question have poor implementations.

Jonathing commented:
Re-assigning to Lex, as I am having difficulties testing this on my setup. Unsealing all events in `RenderTooltipEvent` seems to fix initial crashes on all the modpacks I've tested, but I am still very hesitant to merge this.

LexManos commented:
Ya. After reading the discussion on discord noteing the fact that it breaks atleast one major modpack. Imngoing to say no to this implementation.

A potential backwarss compatible way to address this. At the cost of potentially more memory would be to have an EventListener rebuild its internal list when a child event is encountered for the first time.

PaintNinja committed:
Unseal RenderTooltipEvent and subclasses

Workaround for a bug in the Iceberg mod where it extends these and unintentionally causes the Forge events to be fired twice.
